src/main/daml/Daml/Finance/Instrument/StructuredProduct/Util.daml

```
@@ -0,0 +1,75 @@
+import ContingentClaims.Core.Observation (Observation(..))
+import DA.List (head, last)
+import DA.Optional (isSome)
+import Daml.Finance.Claims.Util.Builders (calculateRatePayment, prepareAndTagClaims)
+import Daml.Finance.Interface.Claims.Types (Deliverable, Observable, TaggedClaim(..))
+import Daml.Finance.Interface.Instrument.Types.FloatingRate (FloatingRate)
+import Daml.Finance.Interface.Types.Date.Calendar (HolidayCalendarData)
+import Daml.Finance.Interface.Types.Date.DayCount (DayCountConventionEnum)
+import Daml.Finance.Interface.Types.Date.Schedule (PeriodicSchedule(..), Schedule)
+import Prelude hiding (and, or, (<=))
+
+type O = Observation Date Decimal Observable
+type C = Claim Date Decimal Deliverable Observable
+
+-- | Find out which schedule periods of scheduleA exist in scheduleB.
+includes : Schedule -> Schedule -> [Bool]
+includes scheduleA scheduleB =
+  map (\a -> isSome $ find (\b -> b.adjustedEndDate == a.adjustedEndDate) scheduleB) scheduleA
+
+-- | Calculate the claims for an auto-callable with a contingent coupon on each payment date
+--   and a redemption amount at maturity (unless auto-called previously).
+createAutoCallableClaims : (Date -> Time) -> Schedule -> Schedule -> PeriodicSchedule
+  -> Bool -> Decimal -> DayCountConventionEnum -> Decimal -> Deliverable -> Optional FloatingRate
+  -> Optional Decimal -> Optional Decimal -> Text -> Decimal -> Decimal -> Decimal -> Decimal
+  -> HolidayCalendarData -> TaggedClaim
+createAutoCallableClaims dateToTime paymentSchedule callableSchedule periodicSchedule
+  useAdjustedDatesForDcf couponRate dayCountConvention notional cashInstrument floatingRate
+  capRate floorRate spot couponBarrier callBarrier finalBarrier putStrike
+  fixingCalendars =
+  let
+    notionalAmount = scale (Const notional) $ one cashInstrument
+
+    principal = notionalAmount
+    initialObservationDate = (.adjustedStartDate) $ head callableSchedule
```

I would throw a nice error message if the callableSchedule is empty